### PR TITLE
Adding a basic search box - phase 1

### PR DESCRIPTION
**Pull request by anathien**
The idea was to add a typeahead-type search box to the header, where the results would directly navigate to the relevant pages (i.e. no result page).

For this to be possible though, a significant refactor was needed: we needed access to the `project` object when rendering the scaffolding of the Project, Module and Symbol pages. For this, I have done the following:
- Extracted the retrieval of project, module and symbol data into custom hooks
- Killed `FetchProject` in favor of the new hooks
- Refactored `ProjectPage`, `ModulePage` and `SymbolPage` to use the new hooks (and a bit more on the sidebar/content side of symbols)

Once that was done I could move on to the searchbar:
- Added a very basic (I mean **VERY** basic) search bar to the right of the header
- Halfway implemented (up till module content) a method that will gather all the searchable "symbols" to populate the search box
- Made the search results themselves links that navigate to the urls of the assets

What this component does:
- Exists
- Provides typeahead functionality up until module content
- Displays the results (and is scrollable)
- Navigates to the requested asset

What it does **NOT**:
- Look nice
- Close when clicked outside
- Work on the index page
- Work on mobile
- Have a placeholder
- Work for all possible indexed stuff

**Please keep in mind this is a premature PR so we can search while we're making the functionality and the component better.** Be gentle please.

How it looks:
image

And as an added benefit, I also made the sidebar go all the way 😛 
image

### CODE REVIEW

File: www/components/SPA/ModulePage.tsx [hunk 1]
```diff
@@ -1,46 +1,63 @@
...
-            </ContentWithSidebar>
-          );
-        }}
-      />
-    </PageLayout>
+    <div>
+      {isLoading && <div>Loading...</div>}
+      {error && (
+        <div>
+          Failed to load <code>${projectJsonUrl}</code>: ${error.message}
```
Comment (zsol) [inline www/components/SPA/ModulePage.tsx:58]: Shouldn't this be
```suggestion
          Failed to load <code>{projectJsonUrl}</code>: {error.message}
```

File: www/components/core/Search.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    <SearchContainer>
+      <input
+        onChange={(event) => {
+          setSearchTerm(event.target.value);
+        }}
+      />
+      {searchTerm && (
+        <SearchResultContainer>
+          {results.length === 0 && "No results found"}
+          {results?.length > 0 &&
```
Comment (zsol) [inline www/components/core/Search.tsx:73]:
```suggestion
          {results.length > 0 &&
```

File: www/hooks/fetchSymbol.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    if (!symbol) {
+      const exportMatch = module.exports.find(
+        (exp) => withoutPrefix(module.name, exp.name) === symbolName
+      );
+      if (exportMatch) {
+        symbol = exportMatch;
+        symbolType = "export";
+      }
```
Comment (zsol) [inline www/hooks/fetchSymbol.ts:72]: Note: this changes the behavior subtly: if there's an export _and_ a func/var/class with the same symbol name, now we're going to prefer the latter. Before this PR we preferred the former - which was the incorrect behavior.
So 👍

File: www/lib/searchDescriptor.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import {
+  mod as generateModuleUrl,
+  project as generateProjectUrl,
+  symbol as generateSymbolUrl,
+} from "./url";
+
+export type SearchDescriptor = {
+  name: string;
+  // TODO: finish this list
+  type?: "project" | "module" | "function" | string;
```
Comment (zsol) [inline www/lib/searchDescriptor.ts:11]:
```suggestion
  type: "project" | "module" | "function" | string;
```

File: www/lib/searchDescriptor.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    return [];
+  }
+
+  const descriptors = [
+    {
+      name: project.name,
+      type: "project",
+      url: generateProjectUrl(project),
+    },
+  ];
```
Comment (zsol) [inline www/lib/searchDescriptor.ts:84]: nit: we won't want the project to be in the search results here
```suggestion
  const descriptors = [];
```

File: www/lib/searchDescriptor.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+  if (project.modules?.length > 0) {
+    generateModuleDescriptors(descriptors, project);
+  }
```
Comment (zsol) [inline www/lib/searchDescriptor.ts:88]:
```suggestion
  generateModuleDescriptors(descriptors, project);
```

File: www/lib/searchDescriptor.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    if (module.classes?.length > 0) {
+      generateSymbolDescriptors(
+        descriptors,
+        project,
+        module,
+        module.classes,
+        "class"
+      );
+    }
```
Comment (zsol) [inline www/lib/searchDescriptor.ts:68]: These fields are guaranteed to exist
```suggestion
    generateSymbolDescriptors(
      descriptors,
      project,
      module,
      module.functions,
      "function"
    );
    generateSymbolDescriptors(
      descriptors,
      project,
      module,
      module.variables,
      "variable"
    );
    generateSymbolDescriptors(
      descriptors,
      project,
      module,
      module.classes,
      "class"
    );
```

File: www/hooks/fetchModule.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  const { project, projectJsonUrl, error, isLoading } =
+    useFetchProject(projectName);
+
+  return {
+    isLoading,
+    error,
+    project,
+    projectJsonUrl,
+    module: project
+      ? project.modules.find((mod) => mod.name === moduleName)
```
Comment (zsol) [inline www/hooks/fetchModule.ts:13]: Is the entire purpose of this hook this `find` call? Is that worth it?

**Review by zsol — APPROVED** [review]

File: www/components/SPA/ModulePage.tsx (hunk 1 above)
Comment (anathien) [reply www/components/SPA/ModulePage.tsx]: Good catch! Fixed in 06b489dcec4b2284ce33b29e8cc5abb674445bdb

File: www/components/core/Search.tsx (hunk 2 above)
Comment (anathien) [reply www/components/core/Search.tsx]: Fair point. Removed in 266a45276f28cd07ae22acfa3c3fde4fb47e78d8

File: www/lib/searchDescriptor.ts (hunk 4 above)
Comment (anathien) [reply www/lib/searchDescriptor.ts]: As discussed offline, adding "export" as a potential value will take care of the case I thought needed the nullable value. Fixed in e132778240ed9f18adcc356f9d7a9b934130c880, also making the possible values explicit.

File: www/lib/searchDescriptor.ts (hunk 5 above)
Comment (anathien) [reply www/lib/searchDescriptor.ts]: Roger that! Removed in 55a020cd5f2141b82f66c50a6c9dfb4a862b4ea5

File: www/lib/searchDescriptor.ts (hunk 6 above)
Comment (anathien) [reply www/lib/searchDescriptor.ts]: Fixed in ab5657b75e4d3986fa7b5843c23d5ba771240ed2

File: www/lib/searchDescriptor.ts (hunk 7 above)
Comment (anathien) [reply www/lib/searchDescriptor.ts]: Cool! Fixed in 4af269e173a0d4133ca09daaad174b0e10f1d951

File: www/hooks/fetchModule.ts (hunk 8 above)
Comment (anathien) [reply www/hooks/fetchModule.ts:13]: Its purpose is to provide a consistent interface for getting all levels of package information. Getting the project and the symbols is done via hooks (with proper logic encapsulated) and thus I feel that providing a hook for the middle tier of information as well is the consistent behavior.

**Pull request merged**